File: gem_shopping.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer;
    map<string, int> search, kindofgem;
    int range = 100000;

    kindofgem.insert({gems[0], 1});
    bool exist = false;
    for (int gi = 1; gi < gems.size(); gi++) 
        kindofgem[gems[gi]]++;

    int start = 0;
    for (int end = 0; end < gems.size(); end++) {
        search[gems[end]]++;

        if (search.size() == kindofgem.size()) {
            while (search.size() == kindofgem.size()) {
                search[gems[start]]--;
                if (search[gems[start]] == 0) search.erase(gems[start]);
                start++;
            }
            if (end - (start-1) < range) {
                range = end - (start-1);
                answer = { start,end+1 };
            }
        }
    }

    return answer;
}

#https://programmers.co.kr/learn/courses/30/lessons/67258
#카카오 2020 인턴십 테스트 : https://tech.kakao.com/2020/07/01/2020-internship-test/
#ref:오영헌

File: gem_shopping.py
def solution(gems):
    gems_dict = {}
    
    l_cur = 0
    r_cur = 0
    l_opt = 0
    r_opt = len(gems)-1
    
    yet = 0 # 탈출조건
    for gem in gems:
        gems_dict[gem] = gems_dict.get(gem,0)
    
    while l_cur < len(gems) and r_cur < len(gems) and l_cur<=r_cur:
        gems_dict[gems[r_cur]]+=1
        yet = 0
        
        for i in gems_dict.values():
            if(i==0):
                yet=1
        
        if(yet):
            r_cur+=1 #오른쪽 커서 이동 (탐색범위 증가)
        else:
            if(r_opt-l_opt>r_cur-l_cur):
                l_opt = l_cur   
                r_opt = r_cur
            gems_dict[gems[l_cur]]-=1
            l_cur+=1 #왼쪽 커서 이동 (탐색범위 감소)
    
    answer = [l_opt+1, r_opt+1]
    return answer

#https://programmers.co.kr/learn/courses/30/lessons/67258
#카카오 2020 인턴십 테스트 : https://tech.kakao.com/2020/07/01/2020-internship-test/
